test/testing/run.spec.ts

```diff
@@ -0,0 +1,709 @@
+      },
+    });
+    expectPostRequest({
+      path: '/results',
+      body: {
+        testExternalId: 'my-test-id',
+        testCaseHash: md5(`12`),
+        testCaseBody: { x: 1, y: 2 },
+        testCaseOutput: '1 + 2 = 3',
+      },
+    });
+    expectPostRequest({
+      path: '/results',
+      body: {
+        testExternalId: 'my-test-id',
+        testCaseHash: md5(`34`),
+        testCaseBody: { x: 3, y: 4 },
+        testCaseOutput: '3 + 4 = 7',
+      },
+    });
+    expectPostRequest({
+      path: '/end',
+      body: {
+        testExternalId: 'my-test-id',
+      },
+    });
+  });
+
+  it('handles async evaluators', async () => {
+    class MyEvaluator extends BaseTestEvaluator<MyTestCase, string> {
+      id = 'my-evaluator';
+
+      async evaluateTestCase(): Promise<Evaluation> {
+        return new Promise((resolve) => {
+          setTimeout(() => {
+            resolve({ score: 0.5 });
+          }, 100);
+        });
+      }
+    }
+
+    await runTestSuite<MyTestCase, string>({
+      id: 'my-test-id',
+      testCases: [
+        { x: 1, y: 2 },
+        { x: 3, y: 4 },
+      ],
+      testCaseHash: ['x', 'y'],
+      evaluators: [new MyEvaluator()],
+      fn: (testCase: MyTestCase) =>
+        `${testCase.x} + ${testCase.y} = ${testCase.x + testCase.y}`,
+    });
+
+    expectNumPosts(6);
+    expectPostRequest({
+      path: '/start',
+      body: {
+        testExternalId: 'my-test-id',
+      },
+    });
+    expectPostRequest({
+      path: '/results',
+      body: {
+        testExternalId: 'my-test-id',
+        testCaseHash: md5(`12`),
+        testCaseBody: { x: 1, y: 2 },
+        testCaseOutput: '1 + 2 = 3',
+      },
+    });
+    expectPostRequest({
+      path: '/results',
+      body: {
+        testExternalId: 'my-test-id',
+        testCaseHash: md5(`34`),
+        testCaseBody: { x: 3, y: 4 },
+        testCaseOutput: '3 + 4 = 7',
+      },
+    });
+    expectPostRequest({
+      path: '/evals',
+      body: {
+        testExternalId: 'my-test-id',
+        testCaseHash: md5(`12`),
+        evaluatorExternalId: 'my-evaluator',
+        score: 0.5,
+      },
+    });
+    expectPostRequest({
+      path: '/evals',
+      body: {
+        testExternalId: 'my-test-id',
+        testCaseHash: md5(`34`),
+        evaluatorExternalId: 'my-evaluator',
+        score: 0.5,
+      },
+    });
+    expectPostRequest({
+      path: '/end',
+      body: {
+        testExternalId: 'my-test-id',
+      },
+    });
+  });
+
+  it('allows specifying a custom hash function', async () => {
+    await runTestSuite<MyTestCase, string>({
+      id: 'my-test-id',
+      testCases: [
+        { x: 1, y: 2 },
+        { x: 3, y: 4 },
+      ],
+      testCaseHash: (testCase: MyTestCase) => `${testCase.x}-${testCase.y}`,
+      evaluators: [],
+      fn: (testCase: MyTestCase) =>
+        `${testCase.x} + ${testCase.y} = ${testCase.x + testCase.y}`,
+    });
+
+    expectNumPosts(4);
```

Comment:
Really like how you define your own expect functions, tests are really easy to follow